[BUG] Clang-tidy: False positive when using `hicpp-signed-bitwise.IgnorePositiveIntegerLiterals`

Consider the following code:

```c
#include "stdint.h"
#include "stdio.h"
#include "inttypes.h"

int main(void) {
    uint32_t u32val = 0;

    printf("%"PRIu32, u32val & 1);
    printf("%"PRIu32, u32val << 2);

    return 0;
}
```

And the following config:

```yaml
# .clang-tidy
---
FormatStyle: file
WarningsAsErrors: '*'
Checks: >
  -*,
  hicpp-signed-bitwise,
CheckOptions:
  hicpp-signed-bitwise.IgnorePositiveIntegerLiterals: true
```

When `clang-tidy` is ran this is the result:

```
test.c:8:32: error: use of a signed integer operand with a binary bitwise operator [hicpp-signed-bitwise,-warnings-as-errors]
    8 |     printf("%"PRIu32, u32val & 1);
      |                              ~ ^
```

The docs for the `IgnorePositiveIntegerLiterals` option state:

> If this option is set to true, the check will not warn on bitwise operations with positive integer literals.

What I believe is happening here is that implicit conversions are not skipped for the RHS operand, thus when performing the `&` op, the literal `1` is implicitly cast from an `int` to an `unsigned int`, and is no longer considered a literal. 

But the same thing doesn't happen for the `<<` operation, because for that operation the operands on both sides do not have to be the same type, thus the literal `2` is not implicitly casted.

Note that implicit casts are skipped for the LHS operand. This can be observed with the following code:

```c
#include "stdint.h"
#include "stdio.h"
#include "inttypes.h"

int main(void) {
    uint8_t u8val = 0;

    printf("%"PRIu8, u8val & 1);
    printf("%"PRIu8, u8val << 2);

    return 0;
}
```

The above code raises no errors when checked with `clang-tidy`, even though `u8val` is implicitly casted to an `int` for both the `&` and `<<` operations.

--- a/clang-tools-extra/clang-tidy/hicpp/SignedBitwiseCheck.cpp
+++ b/clang-tools-extra/clang-tidy/hicpp/SignedBitwiseCheck.cpp
@@ -9,6 +9,7 @@
 #include "SignedBitwiseCheck.h"
 #include "clang/AST/ASTContext.h"
 #include "clang/ASTMatchers/ASTMatchFinder.h"
+#include "clang/ASTMatchers/ASTMatchers.h"
 
 using namespace clang::ast_matchers;
 using namespace clang::ast_matchers::internal;
@@ -29,8 +30,8 @@ void SignedBitwiseCheck::storeOptions(ClangTidyOptions::OptionMap &Opts) {
 void SignedBitwiseCheck::registerMatchers(MatchFinder *Finder) {
   const auto SignedIntegerOperand =
       (IgnorePositiveIntegerLiterals
-           ? expr(ignoringImpCasts(hasType(isSignedInteger())),
-                  unless(integerLiteral()))
+           ? expr(ignoringImpCasts(
+                 allOf(hasType(isSignedInteger()), unless(integerLiteral()))))
            : expr(ignoringImpCasts(hasType(isSignedInteger()))))
           .bind("signed-operand");
 

--- a/clang-tools-extra/docs/ReleaseNotes.rst
+++ b/clang-tools-extra/docs/ReleaseNotes.rst
@@ -247,6 +247,10 @@ Changes in existing checks
 - Improved :doc:`google-runtime-int <clang-tidy/checks/google/runtime-int>`
   check performance through optimizations.
 
+- Improved :doc:`hicpp-signed-bitwise <clang-tidy/checks/hicpp/signed-bitwise>`
+  check by ignoring false positives involving positive integer literals behind
+  implicit casts when `IgnorePositiveIntegerLiterals` is enabled.
+
 - Improved :doc:`hicpp-ignored-remove-result <clang-tidy/checks/hicpp/ignored-remove-result>`
   check by ignoring other functions with same prefixes as the target specific
   functions.

--- a/clang-tools-extra/test/clang-tidy/checkers/hicpp/signed-bitwise-integer-literals.cpp
+++ b/clang-tools-extra/test/clang-tidy/checkers/hicpp/signed-bitwise-integer-literals.cpp
@@ -11,6 +11,7 @@ void examples() {
   // CHECK-MESSAGES: :[[@LINE-1]]:19: warning: use of a signed integer operand with a binary bitwise operator
 
   unsigned URes2 = URes << 1; //Ok
+  unsigned URes3 = URes & 1; //Ok
 
   int IResult;
   IResult = 10 & 2; //Ok
@@ -21,6 +22,8 @@ void examples() {
   IResult = Int << 1;
   // CHECK-MESSAGES: :[[@LINE-1]]:13: warning: use of a signed integer operand with a binary bitwise operator
   IResult = ~0; //Ok
+  IResult = -1 & 1;
+  // CHECK-MESSAGES: :[[@LINE-1]]:13: warning: use of a signed integer operand with a binary bitwise operator [hicpp-signed-bitwise]
 }
 
 enum EnumConstruction {

